crates/red_knot/src/types.rs

```diff
@@ -37,40 +38,137 @@ impl TypeEnvironment {
     }
 }
 
+#[derive(Debug)]
 pub(crate) enum Type {
-    Class(ClassType),
+    // a specific function
     Function(FunctionType),
+    // the set of Python objects with this class in their MRO
+    Instance(ClassType),
+    Union(UnionType),
+    Intersection(IntersectionType),
+    // the dynamic or gradual type: a statically-unknown set of values
+    Any,
+    // the empty set of values
+    Never,
+    // unknown type (no annotation)
+    // equivalent to Any, or to object in strict mode
+    Unknown,
+    // name is not bound to any value
+    Unbound,
+    // TODO protocols, callable types, overloads, generics, type vars
 }
 
 impl Type {
-    pub(crate) fn name(&self) -> &str {
+    pub(crate) fn name(&self) -> Option<&str> {
         match self {
-            Type::Class(inner) => inner.name(),
-            Type::Function(inner) => inner.name(),
+            Type::Instance(inner) => Some(inner.name()),
+            Type::Function(inner) => Some(inner.name()),
+            Type::Union(_) => None,
+            Type::Intersection(_) => None,
+            Type::Any => Some("Any"),
+            Type::Never => Some("Never"),
+            Type::Unknown => Some("Unknown"),
+            Type::Unbound => Some("Unbound"),
+        }
+    }
+
+    fn display(&self, f: &mut std::fmt::Formatter, env: &TypeEnvironment) -> std::fmt::Result {
+        if let Some(name) = self.name() {
+            write!(f, "{}", name)
+        } else {
+            match self {
+                Type::Union(inner) => inner.display(f, env),
+                Type::Intersection(inner) => inner.display(f, env),
+                _ => Err(std::fmt::Error),
+            }
         }
     }
 }
 
+#[derive(Debug)]
 pub(crate) struct ClassType {
     name: Name,
 }
 
 impl ClassType {
-    pub(crate) fn name(&self) -> &str {
+    fn name(&self) -> &str {
         self.name.as_str()
     }
 }
 
+#[derive(Debug)]
 pub(crate) struct FunctionType {
     name: Name,
 }
 
 impl FunctionType {
-    pub(crate) fn name(&self) -> &str {
+    fn name(&self) -> &str {
         self.name.as_str()
     }
 }
 
+#[derive(Debug)]
+pub(crate) struct UnionType {
+    // the union type includes values in any of these types
+    elements: FxHashSet<TypeId>,
+}
+
+impl UnionType {
+    fn display(&self, f: &mut std::fmt::Formatter<'_>, env: &TypeEnvironment) -> std::fmt::Result {
+        write!(f, "(")?;
+        let mut first = true;
+        for elem_id in self.elements.iter() {
+            let ty = env.type_for_id(*elem_id);
+            if first {
+                first = false
+            } else {
+                write!(f, " | ")?;
+            };
+            ty.display(f, env)?;
```

Comment:
```suggestion
            if !first {
                f.write_str(" | ")?;
            };
            
            first = false;
            let ty = env.type_for_id(*elem_id);
            ty.display(f, env)?;
```

Reply:
or you could even move the first out of the loop

```suggestion
	let mut elements = self.elements.iter();
	
	if let Some(first) = elements.next();
		let ty = env.type_for_id(*elem_id);
		ty.display(f, env)?;
	}
	
	for rest in elements {
		let ty = env.type_for_id(*elem_id);
		write!(f, " | {}", ty.display())?;
	}
		
	Ok()
```

Reply:
I like the first suggestion; don't like duplicating most of the contents of the loop